### PR TITLE
Stack parameters

### LINKED ISSUE #16: Generate parameters file based on Parameters section inside the Template
In order to use `aws-cli` with _CloudFormation_ that has a lot of _Parameters_, you need to pass huge amount of command arguments to fill it, or prepare a _JSON_ file with all those values.

We would like to have it available as a command in our tool - we would like to generate a file (sending output to the `stdout`) or command line arguments string only. Values should be read from the standard input (both - interactively or passed as a `stdin` via pipe or something similar).

### PR DESCRIPTION
**Pull request by maxiwoj**
This closes #16.

### CODE REVIEW

File: parameters/parameters.go [hunk 1]
```diff
@@ -0,0 +1,197 @@
...
+func GetAwsParameters(context *context.Context) (parameters []*cloudformation2.Parameter, err error) {
+	var params []*Parameter
+	params, err = GetParameters(context)
+	if err != nil {
+		return
+	}
+	parameters = ParseParameterToAwsCompatible(params)
+	return
+}
+func ParseParameterToAwsCompatible(params []*Parameter) (parameters []*cloudformation2.Parameter) {
```
Comment (afronski) [inline parameters/parameters.go]: Missing empty line?

File: parameters/parameters.go [hunk 2]
```diff
@@ -0,0 +1,197 @@
...
+func parseTemplate(context *context.Context) (res cloudformation.Template, err error) {
+	rawTemplate, err := ioutil.ReadFile(*context.CliArguments.TemplatePath)
+	if err != nil {
+		return
+	}
+
+	myTemplate := template.Template{}
+
+	templateFileExtension := path.Ext(*context.CliArguments.TemplatePath)
+	if templateFileExtension == ".json" {
```
Comment (afronski) [inline parameters/parameters.go]: I believe extension extraction code and decision making can be transferred into common helper, with higher order functions support.

File: parameters/parameters.go [hunk 3]
```diff
@@ -0,0 +1,197 @@
...
+}
+func getAllowedValues(parameterArgument map[string]interface{}) (res []string) {
+	list := parameterArgument["AllowedValues"].([]interface{})
+	for _, val := range list {
+		res = append(res, val.(string))
+	}
+	return
+}
+
+func sliceContains(a string, list []string) bool {
```
Comment (afronski) [inline parameters/parameters.go]: I saw somewhere else similar function - I'd love to see those in helpers.

File: parameters/parameters.go [hunk 4]
```diff
@@ -0,0 +1,197 @@
...
+			return false, err
+		}
+		if !matches {
+			context.Logger.Error("Value '" + parameterValue + "' does not match the required pattern for Parameter " + parameterName)
+			return false, nil
+		}
+	}
+	return true, nil
+}
+func getAllowedValues(parameterArgument map[string]interface{}) (res []string) {
```
Comment (afronski) [inline parameters/parameters.go]: Missing empty line?

File: parameters/parameters.go [hunk 5]
```diff
@@ -0,0 +1,197 @@
...
+	if err != nil {
+		return err
+	}
+	if *context.CliArguments.OutputFilePath != "" {
+		context.Logger.Info("Writing parameters configuration to file: " + *context.CliArguments.OutputFilePath)
+
+		_, err = os.Stat(*context.CliArguments.OutputFilePath)
+		if err == nil {
+			context.Logger.Warning("File " + *context.CliArguments.OutputFilePath + " would be overriten by this action. Do you want to continue? [Y/N]")
+			var ans string
```
Comment (afronski) [inline parameters/parameters.go]: We can go with a full name here. ;)

File: parameters/parameters.go [hunk 6]
```diff
@@ -0,0 +1,197 @@
...
+func GetJSONParameters(context *context.Context) (resultString []byte, err error) {
+	var parameters []*Parameter
+	parameters, err = GetParameters(context)
+	if err != nil {
+		context.Logger.Error(err.Error())
+		return
+	}
+
+	if *context.CliArguments.PrettyPrint {
+		resultString, err = json.MarshalIndent(parameters, "", "    ")
```
Comment (afronski) [inline parameters/parameters.go]: Let's move that into common helper in order to have exact same pretty printing used everywhere.

File: helpers/formathelpers.go [hunk 7]
```diff
@@ -0,0 +1,66 @@
...
+	} else if templateFileExtension == ".yaml" || templateFileExtension == ".yml" {
+		return ParseYAML, nil
+	} else {
+		return nil, errors.New("Invalid template file format.")
+	}
+}
+
+func ParseJSON(templateFile []byte, refTemplate template.Template, logger *logger.Logger) (template cloudformation.Template, err error) {
+
+	err = json.Unmarshal(templateFile, &refTemplate)
```
Comment (jlampar) [inline helpers/formathelpers.go]: Shouldn't there be line and column detection?

```
err = json.Unmarshal(templateFile, &refTemplate)
	if err != nil {
		if syntaxError, isSyntaxError := err.(*json.SyntaxError); isSyntaxError {
			syntaxOffset := int(syntaxError.Offset)
			line, character := lineAndCharacter(string(templateFile), syntaxOffset)
			logger.Error("Syntax error at line " + strconv.Itoa(line) + ", column " + strconv.Itoa(character))
		} else if typeError, isTypeError := err.(*json.UnmarshalTypeError); isTypeError {
			typeOffset := int(typeError.Offset)
			line, character := lineAndCharacter(string(templateFile), typeOffset)
			logger.Error("Type error at line " + strconv.Itoa(line) + ", column " + strconv.Itoa(character))
		}
		return template, err
	}
```

Comment (jlampar) [inline helpers/formathelpers.go]: Shouldn't there be line and column detection?

```
err = json.Unmarshal(templateFile, &refTemplate)
	if err != nil {
		if syntaxError, isSyntaxError := err.(*json.SyntaxError); isSyntaxError {
			syntaxOffset := int(syntaxError.Offset)
			line, character := lineAndCharacter(string(templateFile), syntaxOffset)
			logger.Error("Syntax error at line " + strconv.Itoa(line) + ", column " + strconv.Itoa(character))
		} else if typeError, isTypeError := err.(*json.UnmarshalTypeError); isTypeError {
			typeOffset := int(typeError.Offset)
			line, character := lineAndCharacter(string(templateFile), typeOffset)
			logger.Error("Type error at line " + strconv.Itoa(line) + ", column " + strconv.Itoa(character))
		}
		return template, err
	}
```

Comment (maxiwoj) [reply helpers/formathelpers.go]: The only thing I did was moving the function from one file to another :D 
I can certainly add this, seems like an important functionality.

Comment (maxiwoj) [reply helpers/formathelpers.go]: Oh, I see now that's a change from master, that wasn't included in the release_2_0 branch.

**Pull request merged**